### ISSUE TITLE
gil_scoped_acquire can make unsafe CPython calls without holding the GIL

### ISSUE THREAD

**Issue by nickhutchinson** [pybind/pybind11#1364]
## Issue description

The `gil_scoped_acquire` guard can in some cases require the GIL, because it calls detail::get_internals(), which calls into the Python CAPI if` get_internals` hasn't already been called with the GIL held in the current module.

Reproduced with Pybind 2.2.2, CPython 2.7.13; issue still appears in pybind11 master.

## Reproducible example code

Contrived example:

```
// noddy.cpp
#include <Python.h>
#include <pybind11/pybind11.h>

PyMODINIT_FUNC
initnoddy(void) 
{
    Py_BEGIN_ALLOW_THREADS  // release the GIL
    {
        pybind11::gil_scoped_acquire guard;  // <-- first pybind call in this module
        // Bang!
    }

    Py_END_ALLOW_THREADS  // re-acquire the GIL
}
```

Edit: Original report used the incorrect type for the `guard` variable; have fixed.

**Comment by nickhutchinson** [comment]
My workaround is to force pybind to initialise itself by calling `pybind11::get_shared_data("");` with the GIL already acquired.

**Comment by amosbird** [comment]
@nickhutchinson Hi, this workaround doesn't work for me. I still got segfault when trying to acquire gil in a new thread.

**Comment by wjakob** [comment]
I'm trying to understand why one would ever want to do this. :) Can you explain your use case?

**Comment by nickhutchinson** [comment]
Use case is wanting to make use of pybind11 in an existing Python module that’s not exposed via the PYBIND11_MODULE() mechanism, so gil_scoped_acquire’s hidden assumption that pybind11 has already been initialised for the current module doesn’t necessarily hold.

I’d be happy with a way to explicitly initialise pybind11 for the current module (pybind11::ensure_initialized() ?) and some documentation warning that you need to call it yourself if you’re not using PYBIND11_MODULE(), but I imagine it’d also be possible to tweak the lazy initialisation logic so that it’s not necessary to have an explicit call. 

Nick

**Comment by wjakob** [comment]
I see. This is a little too odd of a use case for us to support in the library. I suggest that you cook up some sort of `ensure_initialized` within your own codebase or modify pybind11 in a branch.

**Comment by saran-t** [comment]
I don't think this is an overly odd use case actually. The "external module" in my case is a collection of C++ functions that may need to call into Python under certain circumstance. This module is a common piece of library code that is depended on by one or more `PYBIND11_MODULE`s, and so itself is compiled into a separate shared library but is not `import`able into Python (since it's at heart a C++ library).

As things stand, `py::gil_scoped_acquire` cannot be treated as a pure function call. This results in hard-to-debug segfaults and is an arguably surprising behaviour. It requires library code to use `PyGILState_Ensure` directly even when it is using `py::` calls everywhere else.

In any case, semantically `py::gil_scoped_acquire` must generically expect to be called in a GIL-released state. The fact that it has a code branch that makes Python calls without GIL should be considered a bug in my view.

**Comment by wjakob** [comment]
Hm, that is a good point. I'll merge the PR in that case, it seems self-contained enough.

**Comment by nickhutchinson** [comment]
While it's mostly a moot point now as I'm no longer in the role where this issue came up, I'd argue that being able to coexist with non-Pybind code is hardly an obscure use case. ;) 

Without this ability, I think pybind is a poor fit for anyone not writing a new Python/C++ module from scratch, as it's unclear which functions are safe to call without triggering segfaults. For example, it means an organisation with a large existing Python/C++ codebase can't migrate it to pybind piecemeal, nor use pybind APIs in Python/C++ modules that need to be exposed using tools like Shiboken or SIP (required for interop with the Qt framework).

**Comment by ingomueller-net** [comment]
For what it's worth, [issue 1738](https://github.com/pybind/pybind11/issues/1738) lists other people having this issue, who might also be happy about a fix...

**Comment by wjakob** [comment]
I merged the PR, so the issue should be fixed as of b60fd233fa423ff56579513efd90820509efb2cc.
